src/transaction.ts

```diff
@@ -0,0 +1,359 @@
+    return this._senderPubKey!
+  }
+
+  /**
+   * Determines if the signature is valid
+   */
+  verifySignature(): boolean {
+    const msgHash = this.hash(false)
+    // All transaction signatures whose s-value is greater than secp256k1n/2 are considered invalid.
+    if (this._common.gteHardfork('homestead') && new BN(this.s).cmp(N_DIV_2) === 1) {
+      return false
+    }
+
+    try {
+      const v = bufferToInt(this.v)
+      const useChainIdWhileRecoveringPubKey =
+        v >= this._chainId * 2 + 35 && this._common.gteHardfork('spuriousDragon')
+      this._senderPubKey = ecrecover(
+        msgHash,
+        v,
+        this.r,
+        this.s,
+        useChainIdWhileRecoveringPubKey ? this._chainId : undefined,
+      )
+    } catch (e) {
+      return false
+    }
+
+    return !!this._senderPubKey
+  }
+
+  /**
+   * sign a transaction with a given private key
+   * @param privateKey - Must be 32 bytes in length
+   */
+  sign(privateKey: Buffer) {
+    const msgHash = this.hash(false)
+    const sig = ecsign(msgHash, privateKey)
+    if (this._chainId > 0) {
+      sig.v += this._chainId * 2 + 8
+    }
+    Object.assign(this, sig)
+  }
+
+  /**
+   * The amount of gas paid for the data in this tx
+   */
+  getDataFee(): BN {
+    const data = this.raw[5]
+    const cost = new BN(0)
+    for (let i = 0; i < data.length; i++) {
+      data[i] === 0
+        ? cost.iaddn(this._common.param('gasPrices', 'txDataZero'))
+        : cost.iaddn(this._common.param('gasPrices', 'txDataNonZero'))
+    }
+    return cost
+  }
+
+  /**
+   * the minimum amount of gas the tx must have (DataFee + TxFee + Creation Fee)
+   */
+  getBaseFee(): BN {
+    const fee = this.getDataFee().iaddn(this._common.param('gasPrices', 'tx'))
+    if (this._common.gteHardfork('homestead') && this.toCreationAddress()) {
+      fee.iaddn(this._common.param('gasPrices', 'txCreation'))
+    }
+    return fee
+  }
+
+  /**
+   * the up front amount that an account must have for this transaction to be valid
+   */
+  getUpfrontCost(): BN {
+    return new BN(this.gasLimit).imul(new BN(this.gasPrice)).iadd(new BN(this.value))
+  }
+
+  /**
+   * Validates the signature and checks to see if it has enough gas.
+   */
+  validate(): boolean
+  validate(stringError: false): boolean
+  validate(stringError: true): string
+  validate(stringError: boolean = false): boolean | string {
+    const errors = []
+    if (!this.verifySignature()) {
+      errors.push('Invalid Signature')
+    }
+
+    if (this.getBaseFee().cmp(new BN(this.gasLimit)) > 0) {
+      errors.push([`gas limit is too low. Need at least ${this.getBaseFee()}`])
+    }
+
+    if (stringError === false) {
+      return errors.length === 0
+    } else {
+      return errors.join(' ')
+    }
+  }
+
+  /**
+   * Returns the rlp encoding of the transaction
+   */
+  serialize(): Buffer {
+    // Note: This never gets executed, defineProperties overwrites it.
```

Comment:
Hm, I think I had also added this to `ethereumjs-account`. But I think you're right that it gets overwritten by `defineProperties`. In that case would there be a reason for keeping it (apart from documentation)?

Reply:
It's needed for TS. You wouldn't be able to call `tx.serialize()` otherwise, as it won't be present in the `Transaction` type.

We may be able to do declare it as a field, something like `serialize!: () => Buffer`. Not sure what's more confusing.

We probably want to do that for `toJSON`. Otherwise, we'll have to duplicate a lot of logic.